attest/eventlog.go

```diff
@@ -241,46 +249,88 @@ func extend(pcr, replay []byte, e rawEvent) ([]byte, Event, error) {
 			hash.Write(b)
 		}
 		hash.Write(digest)
-		return hash.Sum(nil), Event{e.index, e.typ, e.data, digest}, nil
+		return hash.Sum(nil), digest, nil
 	}
-	return nil, Event{}, fmt.Errorf("no event digest matches pcr length: %d", len(pcr))
+	return nil, nil, fmt.Errorf("no event digest matches pcr length: %d", len(pcr))
 }
 
-func replayEvents(rawEvents []rawEvent, pcrs []PCR) ([]Event, error) {
-	events := []Event{}
-	replay := map[int][]byte{}
-	pcrByIndex := map[int][]byte{}
-	for _, pcr := range pcrs {
-		pcrByIndex[pcr.Index] = pcr.Digest
-	}
+// replayPCR replays the event log for a specific PCR, using pcr and
+// event digests with the algorithm in pcr. An error is returned if the
+// replayed values do not match the final PCR digest, or any event tagged
+// with that PCR does not posess an event digest with the specified algorithm.
+func replayPCR(rawEvents []rawEvent, pcr PCR) ([]Event, error) {
+	var (
+		replay    []byte
+		hadEvents bool
+	)
 
+	var outEvents []Event
 	for i, e := range rawEvents {
-		pcrValue, ok := pcrByIndex[e.index]
-		if !ok {
-			// Ignore events for PCRs that weren't included in the quote.
+		if e.index != pcr.Index {
 			continue
 		}
-		replayValue, event, err := extend(pcrValue, replay[e.index], e)
+		hadEvents = true
+
+		replayValue, digest, err := extend(pcr.Digest, replay, e)
 		if err != nil {
-			return nil, fmt.Errorf("replaying event %d: %v", i, err)
+			return nil, fmt.Errorf("replaying event[%d] in PCR[%d]: %v", i, pcr.Index, err)
 		}
-		replay[e.index] = replayValue
-		events = append(events, event)
+		replay = replayValue
+		outEvents = append(outEvents, Event{sequence: e.sequence, Data: e.data, Digest: digest, Index: pcr.Index, Type: e.typ})
+	}
+
+	if hadEvents && !bytes.Equal(replay, pcr.Digest) {
+		return nil, fmt.Errorf("PCR[%d] failed to replay using crypto.Hash(%v)", pcr.Index, pcr.DigestAlg)
```

Comment:
nit: HashAlg already has a Stringer implementation

Reply:
`DigestAlg` is `crypto.Hash`, not `HashAlg`. We can move it to use HashAlg if you want?

Reply:
nah, it's fine as is. plus we just agreed to return a bool instead of an error anyway.